cmd/virt-launcher/virt-launcher.go

```diff
@@ -510,55 +510,37 @@ func ForkAndMonitor(containerDiskDir string) (int, error) {
 		return 1, err
 	}
 
-	exitStatus := make(chan syscall.WaitStatus, 10)
-	sigs := make(chan os.Signal, 10)
-	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD)
-	go func() {
-		for sig := range sigs {
-			switch sig {
-			case syscall.SIGCHLD:
-				var wstatus syscall.WaitStatus
-				wpid, err := syscall.Wait4(-1, &wstatus, syscall.WNOHANG, nil)
-				if err != nil {
-					log.Log.Reason(err).Errorf("Failed to reap process %d", wpid)
-				}
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD)
 
-				// there's a race between cmd.Wait() and syscall.Wait4 when
-				// cleaning up the cmd's pid after it exits. This allows us
-				// to detect the correct exit code regardless of which wait
-				// wins the race.
-				if wpid == cmd.Process.Pid {
-					exitStatus <- wstatus
-				}
+	// wait for virt-launcher and collect the exit code
+	exitCode := 0
+	virtLauncherDone := false
+	for !virtLauncherDone {
```

Comment:
I'm a bit afraid of an infinite loop here. Is it possible that although we signal virt-launcher it will somehow hang? maybe it's reasonable to have some kind of timeout that when exceeds we'll aggressively kill virt-launcher? Does it make sense?

Reply:
This is possible if a bug gets introduced in the virt-launcher (worker) SIGTERM sig handler which makes it hang forever, adding a timeout + SIGKILL sounds like a good idea :+1:

Reply:
I wonder if it makes sense to add it though, cause if this kind of bug gets then 'masked' by this change maybe the CI will not see it anymore ? But I think the change makes sense anyway

Reply:
You have a point. Perhaps we should have a CI test that verifies that the timeout did not exceed here, just to gain indication if something else might be broken (maybe we can run it in periodics)

Reply:
Actually all the test that start a VMI a later delete it are basically testing this (like 'test_id:1651') so I don't think we actually need to add anything.

Reply:
And also when you delete a VMI virt-controller deletes the VMI's pod, which means that kubernetes sends a SIGTERM to the pod, if after the grace period the pod is still running k8s then sends a SIGKILL.

Essentially while I think you made a great suggestion I don't think we should do anything here as kubernetes already does it. wdyt?

Reply:
Well, your words make total sense. With that said, infinite loops are always a bit scary. IMHO we can at least log that we're hanging if such a situation happens so we would have some kind of indication if things go wrong.

But I might be over-thinking this :)

Reply:
Well, we already say "signalling virt-launcher to shut down" when we tell the child virt-launcher to terminate, so if you see that log line and the container still running after a long time I guess you have to worry.

Also keep in mind that if the container is not stopped (meaning sigkilled if we have a bug in virt-launcher TERM handler) and left there hanging forever it means there's also a bug somewhere else (k8s/container runtime/kernel ..) that it's not really our territory (at least to fix).

Reply:
> Well, we already say "signalling virt-launcher to shut down"

Yup, you're right :)

> Also keep in mind that if the container is not stopped (meaning sigkilled if we have a bug in virt-launcher TERM handler) and left there hanging forever it means there's also a bug somewhere else (k8s/container runtime/kernel ..) that it's not really our territory (at least to fix).

That's true, but even if there's a bug that is not our fault we should be aware of where it's coming from. But as you rightfully said, we already log about it, so I guess we're fine :+1: